src/main/java/org/terasology/logic/inventory/InventoryUIClientSystem.java

```diff
@@ -69,4 +91,121 @@ public void onToggleInventory(InventoryButton event, EntityRef entity, ClientCom
+        EntityRef movingItemSlot = playerEntity.getComponent(CharacterComponent.class).movingItem;
+
+        movingItemItem = invManager.getItemInSlot(movingItemSlot, 0);
+        
+        movingItemCount = invManager.getStackSize(movingItemItem);
+
+        EntityRef fromEntity = movingItemSlot;
+        int fromSlot = 0;
+
+        InventoryComponent playerInventory = playerEntity.getComponent(InventoryComponent.class);
+
+        if (movingItemItem != EntityRef.NULL) {
+
+
+
+            if (playerInventory == null) {
+                return;
+            }
+            CharacterComponent characterComponent = playerEntity.getComponent(CharacterComponent.class);
+            if (characterComponent == null) {
+                logger.error("Character entity of player had no character component");
+                return;
+            }
+            int totalSlotCount = playerInventory.itemSlots.size();
+
+            EntityRef interactionTarget = characterComponent.predictedInteractionTarget;
+            InventoryComponent interactionTargetInventory = interactionTarget.getComponent(InventoryComponent.class);
+
+
+            EntityRef targetEntity;
+            List<Integer> toSlots = new ArrayList<>(totalSlotCount);
+            if (fromEntity.equals(playerEntity)) {
+
+                if (interactionTarget.exists() && interactionTargetInventory != null) {
+                    targetEntity = interactionTarget;
+                    toSlots = numbersBetween(0, interactionTargetInventory.itemSlots.size());
+                } else {
+                    targetEntity = playerEntity;
+
+                    toSlots = numbersBetween(0, totalSlotCount);
+
+                }
+            } else {
+                targetEntity = playerEntity;
+                toSlots = numbersBetween(0, totalSlotCount);
+            }
+
+            invManager.moveItemToSlots(getTransferEntity(), fromEntity, fromSlot, targetEntity, toSlots);
+        }
+    }
+
+    @Override
+    public void postAutoSave() {
+        if (movingItemItem != EntityRef.NULL) {
+            EntityRef playerEntity = localPlayer.getCharacterEntity();
+            EntityRef movingItem = playerEntity.getComponent(CharacterComponent.class).movingItem;
+
+            EntityRef targetEntity = movingItem;
+            EntityRef fromEntity = playerEntity;
+
+            int currentSlot = playerEntity.getComponent(InventoryComponent.class).itemSlots.size() - 1;
+
+            while (currentSlot >= 0 && movingItemCount > 0) {
+
+                EntityRef currentItem = invManager.getItemInSlot(playerEntity, currentSlot);
+                int currentItemCount = invManager.getStackSize(currentItem);
+                boolean correctItem = (currentItem == movingItemItem);
+                System.out.println("Current Item: " + currentItemCount);
+                System.out.println("Moving Item Item: " + movingItemCount);
+
+                if (correctItem && movingItemCount <= currentItemCount) {
```

Comment:
This should be changed/simplified as we are doing the same move action in both cases.